modules/apps/object/object-service/src/main/java/com/liferay/object/internal/related/models/SystemObjectMtoMObjectRelatedModelsProviderImpl.java

```diff
@@ -254,6 +257,71 @@ private GroupByStep _getGroupByStep(
 		Column<DynamicObjectRelationshipMappingTable, Long> primaryKeyColumn2 =
 			dynamicObjectRelationshipMappingTable.getPrimaryKeyColumn2();
 
+		ObjectField titleObjectField =
+			_objectFieldLocalService.fetchObjectField(
+				objectDefinition2.getTitleObjectFieldId());
+
+		if ((_localizationTable != null) && titleObjectField.isLocalized()) {
+			return fromStep.from(
+				dynamicObjectDefinitionTable
+			).innerJoinON(
+				dynamicObjectRelationshipMappingTable,
+				primaryKeyColumn2.eq(
+					dynamicObjectDefinitionTable.getPrimaryKeyColumn())
+			).innerJoinON(
+				_localizationTable,
+				dynamicObjectDefinitionTable.getPrimaryKeyColumn(
+				).eq(
+					_localizationTable.getColumn(
+						dynamicObjectDefinitionTable.getPrimaryKeyColumnName())
+				).and(
+					_localizationTable.getColumn(
+						"languageId"
+					).eq(
+						ObjectEntrySearchUtil.getLanguageId()
+					)
+				).and(
+					ObjectEntrySearchUtil.getRelatedModelsPredicate(
+						objectDefinition2, _objectFieldLocalService, search,
+						_localizationTable)
+				)
+			).where(
+				primaryKeyColumn1.eq(
+					primaryKey
+				).and(
+					() -> {
+						Column<?, Long> groupIdColumn = _table.getColumn(
+							"groupId");
+
+						if ((groupIdColumn == null) ||
+							Objects.equals(
+								ObjectDefinitionConstants.SCOPE_COMPANY,
+								objectDefinition1.getScope()) ||
+							Objects.equals(
+								ObjectDefinitionConstants.SCOPE_COMPANY,
+								objectDefinition2.getScope())) {
+
+							return null;
+						}
+
+						return groupIdColumn.eq(groupId);
+					}
+				).and(
+					() -> {
+						Column<?, Long> companyIdColumn = _table.getColumn(
+							"companyId");
+
+						if (companyIdColumn == null) {
+							return null;
+						}
+
+						return companyIdColumn.eq(
+							objectRelationship.getCompanyId());
+					}
+				)
+			);
+		}
```

Comment:
@rijema I noticed that this query is very similar to the one that is retrieved below, can we avoid duplications by creating variables?